aten/src/ATen/core/Generator.cpp

```diff
@@ -0,0 +1,316 @@
+#include "ATen/core/Generator.h"
+
+namespace at {
+
+/*
+* GeneratorState copy and assign constructor implementations
+*/
+GeneratorState::GeneratorState(const GeneratorState& other) {
+  this->device = other.device;
+  this->current_seed = other.current_seed;
+  this->device_type = other.device_type;
+  this->philox_offset_per_thread.exchange(other.philox_offset_per_thread);
+  new (&(this->cpu_engine)) std::mt19937_64(other.cpu_engine);
+}
+
+GeneratorState& GeneratorState::operator=(const GeneratorState& other) {
+  if(&other == this) return *this;
+  this->device = other.device;
+  this->current_seed = other.current_seed;
+  this->device_type = other.device_type;
+  this->philox_offset_per_thread.exchange(other.philox_offset_per_thread);
+  new (&(this->cpu_engine)) std::mt19937_64(other.cpu_engine);
+  return *this;
+}
+
+/*
+* Internal Generator API
+*/
+namespace detail {
+
+// Global generator state and constants
+static std::once_flag cpu_device_flag;
+static GeneratorState* default_gen_state_cpu;
+static std::unique_ptr<Generator> default_gen_cpu;
+
+static int64_t num_gpus = -1;
+static std::once_flag num_gpu_init_flag;
+static std::deque<std::once_flag> cuda_device_flags;
+static std::vector<GeneratorState*> default_gen_states_cuda;
+static std::vector<std::unique_ptr<Generator>> default_gens_cuda;
+
+/* 
+* Populates global values and creates a generator for CPU.
+* Note: the generator on CPU is a 64 bit Mersenne Twister Engine.
+* Warning: this function must only be called once!
+*/
+static void initGlobalCPUGeneratorState(){
+  default_gen_state_cpu = new GeneratorState();
```

Comment:
- Changed the Generator constructor to take a `GeneratorState` object instead of a pointer.
- The pointer in Generator class now is a std::unique_pointer and initializes with c10's make_unique
- The singleton default generators now also utilize c10's make_unique.

Marking as resolved.